api/src/main/java/com/epam/pipeline/manager/cloud/gcp/GCPInstancePriceService.java

```diff
@@ -16,129 +16,154 @@
+                                              final Map<String, String> prefixes) {
+        return objects.stream()
+                .flatMap(machine -> Arrays.stream(GCPResourceType.values())
+                        .filter(machine::isRequired)
+                        .flatMap(type -> requests(machine, type, prefixes)))
                 .collect(Collectors.toList());
     }
 
-    private String termType(final GCPBilling billing) {
-        return billing == GCPBilling.ON_DEMAND ? ON_DEMAND_TERM_TYPE : PREEMPTIBLE_TERM_TYPE;
+    private Stream<GCPResourceRequest> requests(final AbstractGCPObject object,
+                                                final GCPResourceType type,
+                                                final Map<String, String> prefixes) {
+        return Arrays.stream(GCPBilling.values())
+                .map(billing -> Optional.of(object.billingKey(billing, type))
+                        .map(prefixes::get)
+                        .map(prefix -> new GCPResourceRequest(type, billing, prefix, object)))
+                .filter(Optional::isPresent)
+                .map(Optional::get);
     }
 
-    private Double getPrice(final GCPMachine machine,
+    private Double getPrice(final AbstractGCPObject object,
                             final GCPBilling billing,
                             final Set<GCPResourcePrice> prices) {
-        final Map<GCPResourceType, Optional<GCPResourcePrice>> requiredPrices = Arrays.stream(GCPResourceType.values())
-                .filter(type -> type.isRequired(machine))
-                .collect(Collectors.toMap(Function.identity(),
-                    type -> findPrice(prices, type, billing, type.family(machine))));
-        final Optional<GCPResourceType> typeWithMissingPrice = requiredPrices.entrySet()
-                .stream()
-                .filter(entry -> !entry.getValue().isPresent())
-                .map(Map.Entry::getKey)
+        final List<GCPResourceType> requiredTypes = Arrays.stream(GCPResourceType.values())
+                .filter(object::isRequired)
+                .collect(Collectors.toList());
+        final Map<GCPResourceType, GCPResourcePrice> objectPrices = prices.stream()
+                .filter(price -> price.getRequest().getObject().equals(object))
+                .filter(price -> price.getRequest().getBilling().equals(billing))
+                .collect(Collectors.toMap(price -> price.getRequest().getType(), Function.identity()));
+        final Optional<GCPResourceType> typeWithMissingPrice = requiredTypes.stream()
+                .filter(type -> !objectPrices.keySet().contains(type))
                 .findFirst();
         if (typeWithMissingPrice.isPresent()) {
-            log.error(String.format("Price for %s for GCP machine %s wasn't found.", typeWithMissingPrice.get(),
-                    machine.getName()));
+            log.error(String.format("Price for %s with %s billing for GCP object %s wasn't found.",
```

Comment:
You can user `log.error("Message {} ", text) instead of formatting the code manually

Reply:
Done.